signatures/pdf_sign_external_globalsign.go

```diff
@@ -0,0 +1,539 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pdfData, signature, nil
+}
+
+// externalSigner is wrapper for third-party signer,
+// needs to implement function of model.SignatureHandler.
+// in this example, we use GlobalSign as third-party signer.
+type externalSigner struct {
+	// ocsp retrieved during identity request.
+	ocsp      []byte
+	certChain []*x509.Certificate
+
+	// client is third-party api client.
+	// We make client as interface for supporting another third-party signer.
+	client interface{}
+
+	ctx context.Context
+}
+
+// NewGlobalSignPdfSignature wrap externalSigner to model.SignatureHandler.
+// With this, you can implement any third-party client signer into signature handler.
+func NewGlobalSignPdfSignature() (model.SignatureHandler, error) {
+	ctx := context.Background()
+	// Initiate globalsign client.
+	c, err := globalsign.NewClient(apiKey, apiSecret, certFilepath, keyFilepath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &externalSigner{
+		client: c,
+		ctx:    ctx,
+	}, nil
+}
+
+// InitSignature sets the PdfSignature parameters.
+func (es *externalSigner) InitSignature(sig *model.PdfSignature) error {
+	gsClient, ok := es.client.(*globalsign.Client)
+	if !ok {
+		return fmt.Errorf("Not GlobalSign client.")
+	}
+
+	// Request new identification based on signer.
+	identity, err := globalsign.DSSService.DSSGetIdentity(gsClient.DSSService, es.ctx, "GLOBALSIGN TEST ACCOUNT - FOR TESTING PURPOSE ONLY", &globalsign.IdentityRequest{
+		SubjectDn: globalsign.SubjectDn{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// OCSP Response in base64 format.
+	ocsp, err := base64.StdEncoding.DecodeString(identity.OCSP)
+	if err != nil {
+		return fmt.Errorf("invalid ocsp response, err: %v", err)
+	}
+	es.ocsp = ocsp
+
+	// Create certificate chain from signing and CA cert.
+	var certChain []*x509.Certificate
+	issuerCertData := []byte(identity.SigningCert)
+	for len(issuerCertData) != 0 {
+		var block *pem.Block
+		block, issuerCertData = pem.Decode(issuerCertData)
+		if block == nil {
+			break
+		}
+
+		issuer, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return err
+		}
+
+		certChain = append(certChain, issuer)
+	}
+
+	caCertData := []byte(identity.CA)
+	for len(caCertData) != 0 {
+		var block *pem.Block
+		block, caCertData = pem.Decode(caCertData)
+		if block == nil {
+			break
+		}
+
+		issuer, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return err
+		}
+
+		certChain = append(certChain, issuer)
+	}
+	es.certChain = certChain
+
+	// Create PDF array object which will contain the certificate chain data.
+	pdfCerts := core.MakeArray()
+	for _, cert := range certChain {
+		pdfCerts.Append(core.MakeString(string(cert.Raw)))
+	}
+
+	// Append cert to signature.
+	sig.Cert = pdfCerts
+
+	handler := *es
+	sig.Handler = &handler
+	sig.Filter = core.MakeName("Adobe.PPKLite")
+	sig.SubFilter = core.MakeName("adbe.pkcs7.detached")
+	sig.Reference = nil
+
+	// Reserve initial size
```

Comment:
Please remove this comment, as this function actually signs the file.

Reply:
Updated at 15bb7144b66b8891f5c68bd64803bb2d9ee32f54